module/zfs/dsl_scan.c

```diff
@@ -2038,91 +3467,427 @@ dsl_scan_scrub_cb(dsl_pool_t *dp,
 	}
 
 	if (needs_io && !zfs_no_scrub_io) {
-		vdev_t *rvd = spa->spa_root_vdev;
+		dsl_scan_enqueue(dp, bp, zio_flags, zb);
+	} else {
+		count_block(scn, dp->dp_blkstats, bp);
+	}
+
+	/* do not relocate this block */
+	return (0);
+}
+
+static void
+dsl_scan_scrub_done(zio_t *zio)
+{
+	spa_t *spa = zio->io_spa;
+	dsl_scan_io_queue_t *queue = zio->io_private;
+
+	abd_free(zio->io_abd);
+
+	if (queue == NULL) {
+		mutex_enter(&spa->spa_scrub_lock);
+		spa->spa_scrub_inflight--;
+		cv_broadcast(&spa->spa_scrub_io_cv);
+		mutex_exit(&spa->spa_scrub_lock);
+	} else {
+		mutex_enter(&queue->q_vd->vdev_scan_io_queue_lock);
+		ASSERT3U(queue->q_num_inflight_zios, >, 0);
+		queue->q_num_inflight_zios--;
+		cv_broadcast(&queue->q_zio_cv);
+		mutex_exit(&queue->q_vd->vdev_scan_io_queue_lock);
+	}
+
+	if (zio->io_error && (zio->io_error != ECKSUM ||
+	    !(zio->io_flags & ZIO_FLAG_SPECULATIVE))) {
+		atomic_inc_64(&spa->spa_dsl_pool->dp_scan->scn_phys.scn_errors);
+	}
+}
+
+/*
+ * Given a scanning zio's information, executes the zio. The zio need
+ * not necessarily be only sortable, this function simply executes the
+ * zio, no matter what it is. The optional queue argument allows the
+ * caller to specify that they want per top level vdev IO rate limiting
+ * instead of the legacy global limiting.
+ */
+static void
+scan_exec_io(dsl_pool_t *dp, const blkptr_t *bp, int zio_flags,
+    const zbookmark_phys_t *zb, dsl_scan_io_queue_t *queue)
+{
+	spa_t *spa = dp->dp_spa;
+	size_t size = BP_GET_PSIZE(bp);
+	vdev_t *rvd = spa->spa_root_vdev;
+	abd_t *data = abd_alloc_for_io(size, B_FALSE);
+
+	if (queue == NULL) {
 		uint64_t maxinflight = rvd->vdev_children * zfs_top_maxinflight;
 
 		mutex_enter(&spa->spa_scrub_lock);
 		while (spa->spa_scrub_inflight >= maxinflight)
 			cv_wait(&spa->spa_scrub_io_cv, &spa->spa_scrub_lock);
 		spa->spa_scrub_inflight++;
 		mutex_exit(&spa->spa_scrub_lock);
+	} else {
+		uint64_t child_cnt = (queue->q_vd->vdev_children != 0) ?
+		    queue->q_vd->vdev_children : 1;
+		uint64_t maxinflight = zfs_scan_leaf_maxinflight * child_cnt;
+		kmutex_t *q_lock = &queue->q_vd->vdev_scan_io_queue_lock;
+
+		mutex_enter(q_lock);
+		while (queue->q_num_inflight_zios >= maxinflight)
+			cv_wait(&queue->q_zio_cv, q_lock);
+		queue->q_num_inflight_zios++;
```

Comment:
One of the `zio_wait()`s in `dsl_scan_sync()` does this. Specifically, the one in this block:

`else if (scn->scn_is_sorted && scn->scn_bytes_pending != 0) {`